vm/virtualbox/virtualbox.go

```diff
@@ -0,0 +1,312 @@
+
+package virtualbox
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/sys/targets"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func init() {
+	vmimpl.Register("virtualbox", vmimpl.Type{Ctor: ctor})
+}
+
+type Config struct {
+	BaseVM string `json:"vm_name"` // name of the base VM
+	Count  int    `json:"count"`   // number of VMs to run in parallel
+}
+
+type Pool struct {
+	env *vmimpl.Env
+	cfg *Config
+}
+
+type instance struct {
+	cfg        *Config
+	debug      bool
+	baseVM     string
+	vmName     string
+	sshPort    int
+	rpcPort    int
+	sshuser    string
+	sshkey     string
+	timeouts   targets.Timeouts
+	serialPath string
+	closed     chan bool
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, err
+	}
+	if cfg.BaseVM == "" {
+		return nil, fmt.Errorf("config param base_vm is empty")
+	}
+	if cfg.Count < 1 || cfg.Count > 128 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1,128]", cfg.Count)
+	}
+	if _, err := exec.LookPath("VBoxManage"); err != nil {
+		return nil, fmt.Errorf("cannot find VBoxManage")
+	}
+	return &Pool{cfg: cfg, env: env}, nil
+}
+
+func (pool *Pool) Count() int { return pool.cfg.Count }
+
+func (pool *Pool) Create(_ context.Context, workdir string, index int) (vmimpl.Instance, error) {
+	timestamp := strconv.FormatInt(time.Now().UnixNano(), 10)
+	serialPath := filepath.Join(workdir, timestamp, "serial")
+	vmName := fmt.Sprintf("syzkaller_%s", timestamp)
+	inst := &instance{
+		cfg:        pool.cfg,
+		debug:      pool.env.Debug,
+		baseVM:     pool.cfg.BaseVM,
+		vmName:     vmName,
+		sshuser:    pool.env.SSHUser,
+		sshkey:     pool.env.SSHKey,
+		timeouts:   pool.env.Timeouts,
+		serialPath: serialPath,
+		closed:     make(chan bool),
+	}
+	if err := inst.clone(); err != nil {
+		return nil, err
+	}
+	if err := inst.boot(); err != nil {
+		return nil, err
+	}
+	return inst, nil
+}
+
+// We avoid immutable disks and reset the image manually, as this proved more reliable when VMs fail to restart cleanly.
+func (inst *instance) clone() error {
+	if inst.debug {
+		log.Logf(0, "cloning VM %q to %q", inst.baseVM, inst.vmName)
+	}
+	if _, err := osutil.RunCmd(2*time.Minute, "", "VBoxManage", "clonevm", inst.baseVM,
+		"--name", inst.vmName, "--register"); err != nil {
+		if inst.debug {
+			log.Logf(0, "clone failed for VM %q -> %q: %v", inst.baseVM, inst.vmName, err)
+		}
+		return err
+	}
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		if inst.debug {
+			log.Logf(0, "failed to listen on 127.0.0.1:0: %v", err)
+		}
+		return err
+	}
+	inst.sshPort = ln.Addr().(*net.TCPAddr).Port
```

Comment:
It's only to get a free port number? We have `vmimpl.UnusedTCPPort()` specifically for that.

There's still a chance of races though. In `vm/qemu`, we transparently [re-create](https://github.com/google/syzkaller/blob/4e1406b4defac0e2a9d9424c70706f79a7750cf3/vm/qemu/qemu.go#L344) VM in this case, but I think it would still work fine if in the end some error is returned - syzkaller will just call Create() again.